tools/validate_spec.py

```diff
@@ -226,10 +323,67 @@ def __exit__(
             self.file.close()
 
-        if exception_value is None and self.classified_incorrectly != 0:
-            raise ValidationError(
-                f"{self.classified_incorrectly} messages were classified incorrectly"
+        if exception_value is None:
+            err = ValidationError()
+            if self.classified_incorrectly != 0:
+                err.append(f"{self.classified_incorrectly} messages were classified incorrectly")
+            if self.reached_coverage < self.target_coverage:
+                err.append(
+                    f"missed target coverage of {self.target_coverage:.2%}, "
+                    f"reached {self.reached_coverage:.2%}"
+                )
+            if len(err.messages) > 0:
+                raise err
+
+    def write_coverage(self, coverage_info: CoverageInformation) -> None:
+        self.__write_coverage_short(coverage_info)
+        self.__write_coverage_full(coverage_info)
+        self.reached_coverage = coverage_info.total_covered_links / coverage_info.total_links
+        self.target_coverage = coverage_info.target_coverage
+
+    @staticmethod
+    def __write_coverage_short(total_coverage: CoverageInformation) -> None:
+        print("\n")
+        print("-" * 80)
+        print(f"{'RecordFlux Validation Coverage Report' : ^80}")
+        print(f"Directory: {os.getcwd()}")
+        print("-" * 80)
+        print(f"{'File' : <40} {'Links' : >10} {'Used' : >10} {'Coverage' : >15}")
+        for file, _ in total_coverage.spec_files.items():
+            file_links = total_coverage.file_total_links(file)
+            file_covered_links = total_coverage.file_covered_links(file)
+            print(
+                f"{file : <40} {file_links : >10} {file_covered_links : >10} "
+                f"{file_covered_links / file_links :>15.2%}"
             )
+        print("-" * 80)
+        total_links = total_coverage.total_links
+        total_covered_links = total_coverage.total_covered_links
+        print(
+            f"{'TOTAL' : <40} {total_links: >10} {total_covered_links : >10} "
+            f"{total_covered_links / total_links :15.2%}"
+        )
+        print("-" * 80)
+
+    @staticmethod
+    def __write_coverage_full(total_coverage: CoverageInformation) -> None:
+        print("\n")
+        print("=" * 80)
+        print(f"{'Link Coverage' : ^80}")
+        print("=" * 80)
+        for file, _ in total_coverage.spec_files.items():
+            print("\n")
+            print(f"{file : ^80}")
+            print("-" * 80)
+            uncovered_links = total_coverage.file_uncovered_links(file)
+            if len(uncovered_links) == 0:
+                print(f"{'all links covered':^80}")
+                continue
+            for link in uncovered_links:
+                print(
+                    f"{str(link.location) if link.location is not None else '':<17}"
+                    f": missing link {link.source.name:^25} -> {link.target.name:^20}"
```

Comment:
I would rather like to have a "None" instead of nothing for missing locations. Furthermore, there should be no missing link locations in the future.

Reply:
Ok, changed in 815cc77c